File: #6_ZigZag_Conversion/Solution.java
public class Solution {
    public String convert(String s, int numRows) {
        if (numRows == 1) return s;
        if (numRows == 0) return null;
        if (s == null || s.length() == 0) return s;
        int size = 2 * numRows - 2;
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < numRows; i++) {
        	for (int j = i; j < s.length(); j += size) {
        		sb.append(s.charAt(j));
        		if (i != 0 && i != numRows - 1 && j + size - 2 * i < s.length()) {
        			sb.append(s.charAt(j + size - 2 * i));
        		}
        	}
        }
        return sb.toString();
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.convert("0123456789abcd", 4));
    }
}

File: #6_ZigZag_Conversion/Solution.py
class Solution:
    # @param {string} s
    # @param {integer} numRows
    # @return {string}
    def convert(self, s, numRows):
        if not s or numRows <= 1 or len(s) <= numRows:
        	return s
        re = []
        for i in range(numRows):
        	sd = 2 * (numRows - i - 1)
        	su = 2 * i
        	cur = i
    		re.append(s[cur])
        	while True:
        		cur += sd
        		if cur >= len(s):
        			break
        		if sd:
	        		re.append(s[cur])
        		cur += su
        		if cur >= len(s):
        			break
        		if su:
	        		re.append(s[cur])
        return ''.join(re)


sol = Solution()
print sol.convert("PAYPALISHIRING", 3)